File: data_structures/Stack.js
// === STACK ===

// Definition: A last in first out data structure where only the most recently added item can be easily accessed or changed.

// Big O:
    // Lookup: O(n)
    // Pop: O(1)
    // Push: O(1)
    // Peek: O(1)

// IMPLEMENTATION - Array:
class MyStack {
    constructor() {
        this.array = []
    }

    /**
     * Retrieves value of item at the top of the stack.
     * @returns Value at the top of the stack.
     */
    peek() {
        return this.array[this.array.length - 1];
    }

    /**
     * Adds an item to the top of the stack. 
     * @param {*} item Value to be added to the top of the stack.
     * @returns Updated stack. 
     */
    push(item) {
        this.array.push(item);
        return this.array;
    }

    /**
     * Removes an item from the top of the stack. 
     * @returns Updated stack. 
     */
    pop() {
        this.array.pop();
        return this.array;
    }
}

// const stackTest = new MyStack();
// console.log(stackTest.push("Bird"));
// console.log(stackTest.push("Plane"));
// console.log(stackTest.peek());
// console.log(stackTest.push("delete me"));
// console.log(stackTest.peek());
// console.log(stackTest.pop());
// console.log(stackTest);


// IMPLEMENTATION - Linked List:

class Node {
    constructor(value) {
        this.value = value;
        this.next = null;
    }
}

class MyStackTwo {
    constructor() {
        this.top = null;
        this.bottom = null;
        this.length = 0;
    }

    /**
     * Retrieves value of item at the top of the stack.
     * @returns Value at the top of the stack.
     */
    peek() {
        return this.top;
    }

    /**
     * Adds an item to the top of the stack. 
     * @param {*} item Value to be added to the top of the stack.
     * @returns Updated stack. 
     */
    push(item) {
        const newNode = new Node(item);
        if (this.length === 0) {
            this.top = newNode;
            this.bottom = newNode;
        } else {
            const pointer = this.top;
            this.top = newNode;
            this.top.next = pointer;
        }
        this.length++;
        return this;
    }
}

const stackTestTwo = new MyStackTwo();
console.log(stackTestTwo)
console.log(stackTestTwo.push("bird"))
console.log(stackTestTwo.push("plane"))
console.log(stackTestTwo.peek())

File: data_structures/Stack.py
# === STACK ===

# Definition: A last in first out data structure where only the most recently added item can be easily accessed or changed.

# Big O:
#     Lookup: O(n)
#     Pop: O(1)
#     Push: O(1)
#     Peek: O(1)

# IMPLEMENTATION - Array:
from sqlalchemy import null


class MyStack:
    def __init__(self):
        self.stack = []
    
    def __str__(self):
        """Prints the attributes of the class object in a dictionary format."""
        return str(self.__dict__)

    def peek(self):
        """Retrieves value of item at the top of the stack."""
        return self.stack[-1]

    def push(self, item):
        """Adds an item to the top of the stack."""
        self.stack.append(item)
        return self.stack

    def pop(self):
        """Removes an item from the top of the stack."""
        self.stack.pop()
        return self.stack
    
# stack_test = MyStack()
# print(stack_test)
# print(stack_test.push("bird"))
# print(stack_test.push("plane"))
# print(stack_test.peek())
# print(stack_test.push("delete me"))
# print(stack_test)
# stack_test.pop()
# print(stack_test)

# IMPLEMENTATION - LINKED LIST
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None
    
    def __str__(self):
        return str({
            "Value": self.value, 
            "Next": self.next
        })

    def __repr__(self):
        return str({
            "Value": self.value, 
            "Next": self.next
        })
    
class MyStackTwo:
    def __init__(self):
        self.top = None
        self.bottom = None
        self.length = 0

    def __str__(self):
        """Prints the attributes of the class object in a dictionary format."""
        return str(self.__dict__)

    def peek(self):
        """Retrieves value of item at the top of the stack."""
        return self.top
    
    def push(self, item):
        """Adds an item to the top of the stack."""
        new_node = Node(item)
        if self.length == 0:
            self.top = new_node
            self.bottom = new_node
        else:
            pointer = self.top
            self.top = new_node
            self.top.next = pointer
        self.length += 1
        return self

    def pop(self):
        if not self.top:
            return None
        if self.top == self.bottom:
            self.bottom = None
        self.top = self.top.next
        self.length -= 1
        return self

stack_test = MyStackTwo()
print(stack_test)
print(stack_test.push("bird"))
print(stack_test.push("plane"))
print(stack_test.peek())
print(stack_test.push("delete me"))
print(stack_test)
print(stack_test.peek())
stack_test.pop()
print(stack_test)
print(stack_test.peek())